brave/brave-core
# Default search engines by region

Fixes brave/brave-browser#893
Fixes brave/brave-browser#731

Overrides Chromium's set of prepopulated search engines with our own set and regional defaults.
Makes Qwant the default search engine for Germany and France.
Modifies search engine retrieval code to preserve the order of search engines.
Adds unit tests for search engine prepopulate data and util.

Deviates from `Issue 731` in the following ways:
* Yahoo search engine is localized (e.g. in France you'll get Yahoo! France, not Yahoo)
* Yandex search engine is localized
* In Canada, in addition to Yahoo!, Yahoo! Quebec is also listed (keyword: yq)

## Submitter Checklist:


## Automated Test Plan
`npm run test -- brave_unit_tests --filter=BraveTemplateURLPrepopulateDataTest.*`
`npm run test -- brave_unit_tests --filter=BraveTemplateURLServiceUtilTest.*`

## Test Plan:

1. Test correct default search engines list on installation

- Delete any current profile you may have
- Set your locale to U.S.
- Install Brave and run it
- Navigate to chrome://settings and check that under the `Search Engine` section `Search engine used in the address bar` is set to Google.
- Click on the pull-down and check that all engines listed in #731 are present (Note that the order is Google, DuckDuckGo, Qwant, then the rest alphabetically).
- Click on `Manage search engines` and verify that the same engines are listed and have correct `Keyword` associated with them.

2. Test correct default search engines list for CA
- Same steps as in 1, but set your locale to Canada
- Verify that `Yahoo! Quebec` is also present in the list

3. Test correct default search engines list for FR, DE
- Same steps as in 1, but set your locale to France (or Germany)
- Verify that Qwant is listed on top and is set as the default search engine
- Also, verify that Yahoo! is listed as localized version (`Yahoo! France` or `Yahoo! Deutschland`)

4. Test that keyword search is working correctly
- Open a new tab
- In the omnibox type one of the `keywords` (e.g. `:g`) and press Tab
- Omnibox should display `Search XXX` where XXX is the search engine associated with the typed keyword

5. Test that the order of engines is preserved when default is changed
- Navigate to `chrome://settings/searchEngines`
- Click on `...` of an engine that isn't the default and select `Make default`
- Verify that the engine is selected as default (bold)
- Verify that the order of engines on the page has not changed

6. Test removing search engine
- Navigate to `chrome://settings/searchEngines`
- Click on `...` of an engine that isn't the default and select `Remove from list`
- Verify that the engine has been removed from the list
- Verify that the order of engines on the page has not changed

7. Test resetting settings
- Perform 5 or 6
- Navigate to `chrome://settings/resetProfileSettings?origin=userclick`
- In the `Reset settings`  pop-up click `Reset settings`
- Navigate to `chrome://settings/searchEngines` and verify that the list has been restored to the original state (removed items are back in the list, default is reset)

8. Test that the engines order is preserved on second run
- Repeat test 1.
- Exit Brave
- Start Brave again and navigate to `chrome://settings/searchEngines`
- Verify that the search engines remain in the correct order

## Reviewer Checklist:
